src/ray/raylet/raylet_client.cc

```diff
@@ -18,228 +18,72 @@
-
+  std::unique_ptr<stream_protocol::socket> _socket;
+  auto status = ray::UnixSocketConnect(raylet_socket, main_service_, _socket);
+  RAY_CHECK_OK_PREPEND(status,
+                       "Unable to connect to raylet socket " + raylet_socket + ".");
+  conn_ = ray::LocalServerConnection::Create(std::move(*_socket.release()));
+  conn_->SetClientID(client_id);
   flatbuffers::FlatBufferBuilder fbb;
   auto message = ray::protocol::CreateRegisterClientRequest(
       fbb, is_worker, to_flatbuf(fbb, client_id), getpid(), to_flatbuf(fbb, driver_id),
       language);
   fbb.Finish(message);
   // Register the process ID with the raylet.
   // NOTE(swang): If raylet exits and we are registered as a worker, we will get killed.
-  auto status = conn_->WriteMessage(MessageType::RegisterClientRequest, &fbb);
+  status = conn_->WriteMessage(MessageType::RegisterClientRequest, fbb.GetSize(),
+                               fbb.GetBufferPointer());
   RAY_CHECK_OK_PREPEND(status, "[RayletClient] Unable to register worker with raylet.");
 }
 
+ray::Status RayletClient::Disconnect() {
+  flatbuffers::FlatBufferBuilder fbb;
+  auto message = ray::protocol::CreateDisconnectClient(fbb);
+  fbb.Finish(message);
+  auto status = conn_->WriteMessage(MessageType::IntentionalDisconnectClient,
+                                    fbb.GetSize(), fbb.GetBufferPointer());
+  // Don't be too strict for disconnection errors.
+  // Just create logs and prevent it from crash.
+  if (!status.ok()) {
+    RAY_LOG(ERROR) << status.ToString()
+                   << " [RayletClient] Failed to disconnect from raylet.";
+  }
+  return ray::Status::OK();
+}
+
 ray::Status RayletClient::SubmitTask(const std::vector<ObjectID> &execution_dependencies,
                                      const ray::raylet::TaskSpecification &task_spec) {
   flatbuffers::FlatBufferBuilder fbb;
   auto execution_dependencies_message = to_flatbuf(fbb, execution_dependencies);
   auto message = ray::protocol::CreateSubmitTaskRequest(
       fbb, execution_dependencies_message, task_spec.ToFlatbuffer(fbb));
   fbb.Finish(message);
-  return conn_->WriteMessage(MessageType::SubmitTask, &fbb);
+  return conn_->WriteMessage(MessageType::SubmitTask, fbb.GetSize(),
+                             fbb.GetBufferPointer());
 }
 
 ray::Status RayletClient::GetTask(
     std::unique_ptr<ray::raylet::TaskSpecification> *task_spec) {
-  std::unique_ptr<uint8_t[]> reply;
   // Receive a task from the raylet. This will block until the local
   // scheduler gives this client a task.
-  auto status =
-      conn_->AtomicRequestReply(MessageType::GetTask, MessageType::ExecuteTask, reply);
-  if (!status.ok()) return status;
+  auto status = conn_->WriteMessage(MessageType::GetTask, 0, nullptr);
+  if (!status.ok()) {
+    return status;
+  }
+  std::unique_ptr<uint8_t[]> reply;
+  status = conn_->ReadMessage(MessageType::ExecuteTask, reply);
+  if (!status.ok()) {
+    return status;
```

Comment:
This change looks to be a breaking one ... that multiple threads sharing the same raylet client could read incorrect replies with this... could you kindly explain the purpose of this change?

Reply:
Similar for the wait function below

Reply:
Thanks for your comment. I know it is a breaking change because I haven't add back the locks. In the original code, data reading is not thread-safe. I guess it is because when you are waiting for a task using `get_task`, another thread could be also updating profiling events. We could definitely do it in a better way by using a more complex mutex design. However, this PR suffers from some unknown memory leaks currently (not caused by data race), so I decide to put the locks aside for a while and figure out what happens.

Reply:
Thanks for the explanation. Sometimes sharing raylet client between different threads can be problematic, and we met some issues in our projects and we ended up adding a different type of client and use that in some cases. Here's an [on-going PR #3889](https://github.com/ray-project/ray/pull/3889) for this.